### PR TITLE
Feature/storage/jumbo blob

### PR DESCRIPTION
**Pull request by jaschrep-msft**
Support for Jumbo Blobs

Raised limits of Put Blob and Put Block to 5000MB and 4000MB respectively.
Replaced one-array-per-block strategy in buffered parallel upload with multi-array buffered streams, allowing buffered blocks to be greater than maxint in size.

To be added in a future PR: genericizing this work into common to support Jumbo DataLake Files.

### CODE REVIEW

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:75]: Let's put one parameter per line.  We have a convention to put all parameters on 1 line, or if there are more than ~3, 1 per line.

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:89]: Consider expanding this statement to multiple lines for readability.

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:213]: Probably want to remove the Console.WriteLine()

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:209]: Can we expand this statement for readability?

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:222]: Consider adding a comment describing this method.

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:66]: Nit: we generally document parameters on 3 lines:
```
///<param name="stream">
/// Stream to buffer from.
///</param>
```

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:80]: Could we add comments for each variable on lines 80-88?

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:31]: Consider re-naming to something along the lines of DataLength.

**Comment by seanmcc-msft** [comment]
@tg-msft and @kasobol-msft, can you take a look?

**Comment by seanmcc-msft** [comment]
Run ```sdk\storage\generate.ps1``` and ```eng\scripts\Export-API.ps1``` to pass the analyze step.  Hold off until @amnguye's PR is in - https://github.com/Azure/azure-sdk-for-net/pull/11409.  There are swagger changes in there that aren't present in the 73 base branch.

Also, it looks like that issue you were talking about is repoing in the net461 tests.

**Comment by kasobol-msft** [comment]
> Run `sdk\storage\generate.ps1` and `eng\scripts\Export-API.ps1` to pass the analyze step. Hold off until @amnguye's PR is in - #11409. There are swagger changes in there that aren't present in the 73 base branch.
> 
> Also, it looks like that issue you were talking about is repoing in the net461 tests.

That's bad (the later part). I wonder if .NET has separate perf test suite being built like Java SDK. If so then maybe we should put these tests there. If not the we probably want to skip this test for 32bits processes somehow.

@jaschrep-msft Other than that could you please check if having jumbo blob tests in CI build is making them run significantly longer? If yes, then please consider marking them "live only".

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:89]: Even better if you could do this https://refactoring.guru/decompose-conditional by either extracting named method OR creating inline bool variable that would give some name to the conditional statement.

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:92]: When I tried to do this in Java I got "API compatibility error". I guess it's not happening here because we alias constant via property. Java error made me think that we probably shouldn't be doing that - let existing code do the same thing in case someone is using this value in some calculations.

File: sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploadExtensions.cs
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploadExtensions.cs:77]: Should this be defined inside PooledMemoryStream?

File: sdk/storage/Azure.Storage.Common/tests/PooledMemoryStreamTests.cs
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Common/tests/PooledMemoryStreamTests.cs:15]: Can we have some "live" tests that actually try to push 4-8GB blocks E2E? So that we catch potential problems in layers on the top of PooledMemoryStream? (i.e. validations, retaining references preventing GC from collecting garbage).

**Review by kasobol-msft — COMMENTED** [review]
Datalake Append is now accepting 4000MB as well. Did we bump that threshold as well?

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:213]: oops

**Comment by jaschrep-msft** [comment]
@kasobol-msft Data lake will be in a separate PR, as called out in the description. The majority of that work is going to be genericizing the current work to apply to both blob and data lake, and I felt that would clutter this PR too much.

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:92]: I don't understand the benefit to leaving this value back at 256 MB.

**Comment by kasobol-msft** [comment]
> @kasobol-msft Data lake will be in a separate PR, as called out in the description. The majority of that work is going to be genericizing the current work to apply to both blob and data lake, and I felt that would clutter this PR too much.

@jaschrep-msft My bad. I should have read description first :-)

File: sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploadExtensions.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploadExtensions.cs:77]: We can introduce a default, and let our code not specify, but I don't think a value like this should be hard-coded into the stream class.

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs
Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:92]: It's more of - what can break if we move publicly available "constant" that someone could take dependency on. Someone may come and say that after upgrade their throughput have changed without them making any change on their side.

File: sdk/storage/Azure.Storage.Common/tests/PooledMemoryStreamTests.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/tests/PooledMemoryStreamTests.cs:15]: Is this not covered by additions to PartitionedUploaderTests where we pump in 2 blocks each maxint plus change in size? Or are you asking I add some checks to that test where I validate every rented array was returned?

Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Common/tests/PooledMemoryStreamTests.cs:15]: The PartitionedUploaderTests focus on single component that's being used in larger process of uploading blobs. It's good that we focus our coverage here.

What I'm saying is that we should have a test or two testing functionality of "jumbo blob" end to end. (I've found few surprises outside of "PooledMemoryStream" equivalent there by doing that).

Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Common/tests/PooledMemoryStreamTests.cs:15]: Well. Sorry. I think I confused Uploader with PooledMemoryStream. You're absolutely right! Please disregard this conversation.

File: sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploadExtensions.cs
Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploadExtensions.cs:77]: If we don't expect to customize that between blob and datalake then I'd rather introduce default. Less code is better.

**Comment by jaschrep-msft** [comment]
> Other than that could you please check if having jumbo blob tests in CI build is making them run significantly longer? If yes, then please consider marking them "live only".

This adds 4 min across all tests. I can mark those as live only, though that's a little weird since they never actually hit the network. Do we prefer to mark them as explicit?

**Comment by seanmcc-msft** [comment]
You can merge the 73 base branch into your feature branch, I fixed the blocking issues.

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:93]: If that is our mindset then I don't think we should expose a new one, keeping all our constants internal and marking this old one `EditorBrowsable.Never`. If we can't change a value that is designed to communicate the current max value, which we know for certain will change again in the future, then we shouldn't be making publicly available values for the current max that we have to keep hiding and exposing a new one for.

Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:93]: I got stopped by static analysis in Java when I tried to do the same there - it told me it's breaking API change. Therefore I'd check .NET SDK guidelines related to this change.

On the other hand - if we can't do this in Java - should we in .NET?

And then... Python doesn't validate that at all...

@tg-msft could you chime in here?

File: sdk/storage/Azure.Storage.Blobs/tests/PartitionedUploaderTests.cs
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/PartitionedUploaderTests.cs:249]: What does `Explicit` do ?

File: sdk/storage/Azure.Storage.Common/src/Shared/Constants.cs
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/Constants.cs:166]: comment is inaccurate.

**Review by kasobol-msft — COMMENTED** [review]
Overall it's looking good. I'm just waiting to see how "max integer" change discussion goes.

File: sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploadExtensions.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploadExtensions.cs:60]: Could we add 1-line comment here explaining this calculation?

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:108]: Super nit: can we put these comments on the line above?  Same for the rest of the comments in this file.

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:185]: Another super-nit comment thing - let's remove ```nonsensical``` and ```useless``` in this and the following comment.

**Review by seanmcc-msft — COMMENTED** [review]
Looking good.  How did you resolve the CI memory issues?

File: sdk/storage/Azure.Storage.Blobs/tests/PartitionedUploaderTests.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/PartitionedUploaderTests.cs:249]: Explicit means that the test must be opted into in some form. According to the nunit website: `The test or fixture will be run if it is selected in the gui, if its name is specified on the console runner command line as the fixture to run, or if it is included by use of a Category filter.`. Note that selecting an enclosing menu in the VS test explorer (e.g. selecting PartitionedUploaderTests or Azure.Storage.Blobs.Test) counts as "selected in the gui," but stating a similar scope on the console does *not* count as "name is specified on the console runner command line as the fixture to run." This means we as devs will likely run the test in our local development cycle but CI will not run this test.

**Comment by jaschrep-msft** [comment]
@seanmcc-msft I marked the huge tests as explicit. They are not actually running in CI, as they are not just memory hogs but also take multiple minutes. This is generally what we do for any "large blob" tests we have as well.

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:93]: We made these properties in .NET instead of constants specifically so we could version them when this happened.  I think we should make this change for .NET even if we can't for Java _(but start using properties for new constants in Java as well so we'll have that wiggle room if we ever need it)_.

This is one of those things that we should really consider checking the version for and return 256 MB for everything less than the current and `int.MaxValue` from here on out.  The original suggestion was to only add version checks when there was a potential for data corruption or loss, but anyone doing a check on this constant to upload with the old service version will have a very unpleasant time.  It would be as as easy as `public virtual int BlockBlobMaxUploadBlobBytes => Version < BlobClientOptions.ServiceVersion.Vxyz ? 256 * Constants.MB : int.MaxValue;`

File: sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:916]: For the PartitionedUploader unit tests.

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:19]: This PR does not address issues of holding everything in memory, or any other optimizations. This simply allows users with good enough hardware to not be limited by our SDK implementation.

Ultimately, we shouldn't stop people who have an unseekable stream of the size where this becomes necessary, want parallel network connections for their upload, and have the hardware to support such a task. This stream as a replacement for single arrays is designed to allow for that, and is what my team considers bare minimum for supporting the next service release.

A future PR for optimizations on uploads will include:
- pure streaming when parallelism is set to 1
- optimizations for when files are used instead of raw streams

Comment (tg-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:19]: > This simply allows users with good enough hardware to not be limited by our SDK implementation.

I don't understand this part of the comment.  Could you elaborate on that?

> Ultimately, we shouldn't stop people who have an unseekable stream of the size where this becomes necessary, want parallel network connections for their upload, and have the hardware to support such a task. This stream as a replacement for single arrays is designed to allow for that, and is what my team considers bare minimum for supporting the next service release.

I'm confused by this part as well, but maybe answering the first part will make this clear too.

> pure streaming when parallelism is set to 1

What does pure streaming mean?

File: sdk/storage/Azure.Storage.Blobs/tests/PartitionedUploaderTests.cs
Comment (pakrym) [reply sdk/storage/Azure.Storage.Blobs/tests/PartitionedUploaderTests.cs:249]: Is there a reason not to run the test by default?

File: sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/PooledMemoryStream.cs:19]: Our old limitation used a single array to store a whole block. Once you break maxint in size, this no longer works (yes technically the api allows for longs but net461 doesn't and also the arraypool api doesn't). MemoryStream also doesn't satisfy our needs because it doesn't hook up to an array pool for optimizations and doesn't actually support long amounts of data either. Interestingly, I found a microsoft published arraypool backed stream, but it also doesn't support long values. So this stream, as it uses multiple arrays, allows us to store these massive blocks in memory should it be necessary.

The argument in the second part is that some people have the hardware to support these massive memory blocks. Our code, using one array per block, didn't allow them to use jumbo blob's full capabilities even though they had the hardware to run it.

By pure streaming I mean not buffering an entire block in memory before uploading it, as the whole point of buffering the entire block is to start uploading it while we begin reading the next block. If we're uploading blocks sequentially, it makes no sense to buffer the entire block beforehand.

File: sdk/storage/Azure.Storage.Blobs/tests/PartitionedUploaderTests.cs
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/PartitionedUploaderTests.cs:249]: It uses >4GB of memory and is on the time scale of minutes not milliseconds. Notably, nunit net461 tests run on x86 by default, while netcoreapp tests run on x64 by default. Locally, I have to explicitly specify x64 or the test will have an out of memory exception. I don't believe I have access to alter CI to run x64, and it's probably not a switch to throw without thinking.

**Review by kasobol-msft — APPROVED** [review]

**Review by seanmcc-msft — APPROVED** [review]
It would be great if @tg-msft could take one more look.

**Comment by jaschrep-msft** [comment]
> 
> 
> It would be great if @tg-msft could take one more look.

I'd also like explicit confirmation that the API changes to transfer options and block blob properties are okay, or feedback as to what needs to shift for such an approval.

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:1265]: The convention in .NET is `Foo` and `LongFoo` but you've got `FooLong` here.  See [Array.LongLength](https://docs.microsoft.com/en-us/dotnet/api/system.array.longlength?view=netcore-3.1) for comparison.  We're hiding the old `Foo` name so that gives  us a little more flexibility.  Maybe we call it `BlockBlobMaxStageBlockLongBytes` so the convention is still consistent but it'll still be very obvious to customers who need to change?

**Review by tg-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/src/PartitionedDownloader.cs
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/src/PartitionedDownloader.cs]: I was actually okay with `Size` for these if you liked that better than `LongLength`.  I thought it was a really good alternative to `Length` that still made sense to users.  I only wanted to explicitly fix the `Long` suffixed properties.

Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/src/PartitionedDownloader.cs]: Okay I'll set that back.

**Review by tg-msft — APPROVED** [review]
Looks great.

**Pull request merged**
